--- /usr/lib/python2.7/site-packages/nova/network/model.py.orig	2017-05-29 18:50:11.970846546 +0000
+++ /usr/lib/python2.7/site-packages/nova/network/model.py	2017-05-29 18:50:32.310797250 +0000
@@ -51,6 +51,7 @@
 # class
 VIF_DETAILS_PORT_FILTER = 'port_filter'
 VIF_DETAILS_OVS_HYBRID_PLUG = 'ovs_hybrid_plug'
+VIF_DETAILS_VHOSTUSER_VROUTER_PLUG = 'vhostuser_vrouter_plug'
 VIF_DETAILS_PHYSICAL_NETWORK = 'physical_network'
 VIF_DETAILS_BRIDGE_NAME = 'bridge_name'
 
