--- ./usr/lib/python2.7/site-packages/nova/network/os_vif_util.py.orig	2017-05-29 18:41:32.260567105 +0000
+++ ./usr/lib/python2.7/site-packages/nova/network/os_vif_util.py	2017-05-29 18:44:39.787924678 +0000
@@ -300,6 +300,19 @@
             raise exception.VifDetailsMissingVhostuserSockPath(
                 vif_id=vif['id'])
         return obj
+    elif vif['details'].get(model.VIF_DETAILS_VHOSTUSER_VROUTER_PLUG, False):
+        vif_details = vif['details']
+        mode = vif_details.get(model.VIF_DETAILS_VHOSTUSER_MODE,
+                               'server')
+        sock_path = vif_details.get(model.VIF_DETAILS_VHOSTUSER_SOCKET)
+        if sock_path is None:
+            raise exception.VifDetailsMissingVhostuserSockPath(
+                                                        vif_id=vif['id'])
+        obj = _get_vif_instance(vif, objects.vif.VIFVHostUser,
+                                mode=mode,
+                                path=sock_path,
+                                plugin="vrouter", vif_name=_get_vif_name(vif))
+        return obj
     else:
         raise NotImplementedError()
 
